feat: test summary

kyu08 opened the pull request:
Resolves #394

This is an implementation of #394. 
This implementation adds output like this.

| Case | Image |
|--------|--------|
|When all tests passed| <img width="559" alt="image" src="https://github.com/zoncoen/scenarigo/assets/49891479/6b783696-f317-44a6-901f-ce21752fd260">|
|When some tests failed|<img width="771" alt="image" src="https://github.com/zoncoen/scenarigo/assets/49891479/adbad2a4-137d-447e-8e8d-93d0adbc61e9">|

If this feature request is acceptable, I will polish up this PR on these points.
- Make this feature configurable.
- If `colored: true` , outputs with some colors.
- Add some unit tests.
- Add feature description to README.

Thank you.

zoncoen commented:
Thank you for your proposal!
As you suggested, it should be configurable. Please add an option to `OutputConfig`.

https://github.com/zoncoen/scenarigo/blob/ffdb628865654047c3e26b05c14d952cbdb60263/schema/config.go#L57-L61

kyu08 commented:
I will add few changes to this PR.
After I finished it I will reopen this PR.
Please review these changes after that. 🙏

zoncoen commented on schema/config.go:
```diff
@@ -55,9 +55,10 @@ type YTTConfig struct {
-	Colored *bool        `yaml:"colored,omitempty"`
-	Report  ReportConfig `yaml:"report,omitempty"`
+	Verbose   bool         `yaml:"verbose,omitempty"`
+	NoSummary bool         `yaml:"noSummary,omitempty"`
```
Please insert `Summary` field into the line below `Colored`.

kyu08 replied:
Fixed in cdf2703b061050efce6c8f45f82e5112b1b0da07

kyu08 commented on cmd/scenarigo/cmd/config/default.scenarigo.yaml:
```diff
@@ -9,6 +9,7 @@ pluginDirectory: ./gen      # Specify the root directory of plugins.
 
 output:
   verbose: false   # Enable verbose output.
+  noSummary: false # Disable show summary.
```
Fixed in cdf2703b061050efce6c8f45f82e5112b1b0da07.

kyu08 commented on runner.go:
```diff
@@ -40,6 +40,8 @@ type Runner struct {
 	inputConfig     schema.InputConfig
 	reportConfig    schema.ReportConfig
+	testSummary     testSummary
+	w               io.Writer
```
Thank you for letting me know the difference of responsibilities of Runner and `reporter`.

This point was fixed in 8a0294b673ba5a612edd71a67f8e63468e7b5e32.

kyu08 commented:
## TODO
- [ ] Fix CI errors.
- [ ] Add unit tests.
- [ ] Update README
    - [ ] Update code snippet generated by `scenarigo init` in `## Setup`

kyu08 commented:
@zoncoen
I'm done with TODOs and the changes you pointed.
This is a summary of changes after your first review.

- implement `PrintSummary` as `reporter`'s method(ce5ca6c)
- fix existing tests's errors caused by my changes (1fa298c, 5e91e6d)
- update readme (bb761bb)
- change `testSummary#add`'s argument(`r reporter.Reporter` to `testResultString string`) for the testability (6d2ac7f)
- add unit tests (a3713ee, b9375b0, f2d457b)
- format(5f31d8e)

Could you review again please?

zoncoen reviewed (changes requested):
Please implement the logic to print a summary in the `reporter` package (not `scenarigo` package). I think it is enough like the following code.

```func run(cmd *cobra.Command, args []string) error {
	// ...
	if cfg.Output.Summary {
		reporterOpts = append(reporterOpts, reporter.WithTestSummary())
	}

	var reportErr error
	success := reporter.Run(
		func(rptr reporter.Reporter) {
			r.Run(context.New(rptr))
			reportErr = r.CreateTestReport(rptr)
		},
		reporterOpts...,
	)
	// ...
}```

https://github.com/zoncoen/scenarigo/blob/3f72845f973421774eecda4a60f222482e8ae5e2/cmd/scenarigo/cmd/run.go#L34

```func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPolicy) bool {
	// ...
	if r.isRoot() {
		printReport(child)  // This is 
		if r.context.summary {
			printSummary()
		}
	}
	return !child.Failed()
}```

https://github.com/zoncoen/scenarigo/blob/3f72845f973421774eecda4a60f222482e8ae5e2/reporter/reporter.go#L265

kyu08 commented:
Thank you for your feed back!

> Please implement the logic to print a summary in the reporter package (not scenarigo package). I think it is enough like the following code.

I'm sorry for misunderstood your advice. I'll fix it.

Also, thank you for showing me code examples kindly.
I tried to print `test message` in a way like your second code snippet like this.
```go
func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPolicy) bool {
        // ...
	if r.isRoot() {
		printReport(child)
		r.context.printf("%s\n", "\ntest message\n")
	}
	return !child.Failed()
}
```

But it seems like `test message` was showed after every test file execution like below.
<img width="828" alt="image" src="https://github.com/zoncoen/scenarigo/assets/49891479/9beeb620-66d8-4f4a-ae68-9ad3b1ee683b">

How about call `printSummary` [here](https://github.com/kyu08/scenarigo/blob/f2d457bb2a3f96916ca463c468e7dc29ab52fcfd/reporter/reporter.go#L56) and  call `printSummary#add` [here](https://github.com/kyu08/scenarigo/blob/f2d457bb2a3f96916ca463c468e7dc29ab52fcfd/reporter/reporter.go#L284)?

When I change to call `r.context.printf` in `Run` func like below.
```go
// reporter/reporter.go
func Run(f func(r Reporter), opts ...Option) bool {
	r := run(f, opts...)
      	r.context.printf("test message called in Run func\n")
	return !r.Failed()
}
```
Then `test message` was shown just once properly.
<img width="696" alt="image" src="https://github.com/zoncoen/scenarigo/assets/49891479/5bb2b52c-2ecd-4f1f-b2e3-e61199e6d6b7">

zoncoen commented:
Ah~ you're right. It's OK to print a summary in `Run` func 👍

kyu08 commented:
I fixed to  implement test summary logic in the `reporter` package.
And I added some minor changes. Please check it out.

### About the naming `summary` versus `testSummary` c138756506be77d13e8650f6991937d8e410952c
I used the naming `summary` in the context of configuration, and I used the naming `testSummary` in other contexts.
Because I feel like it is natural that the naming `summary` means this test summary feature in the config context, on the other hand it seems unnatural in the context of whole code base.

If you have any thought, please give me an idea🙏

zoncoen commented on reporter/reporter.go:
```diff
@@ -276,6 +288,7 @@ func (r *reporter) runWithRetry(name string, f func(t Reporter), policy RetryPol
 	r.appendChildren(child)
 	if r.isRoot() {
 		printReport(child)
+		child.context.testSummary.append(name, TestResultString(child))
```
I don't want to keep any data if the summary feature is disabled. How about like below?

```go
func (s *testSummary) append(testFileRelPath string, r Reporter) {
	if s == nil {
		return // disalbed
	}
	testResultString := TestResultString(r)
	// ...
}
```

kyu08 replied:
Fixed in aca2ad33f985c60f2d3daa30938fd1ebe9e42bf4.

zoncoen commented on reporter/test_summary.go:
```diff
@@ -0,0 +1,98 @@
+		s.skipped = append(s.skipped, testFileRelPath)
+	default: // Do nothing
+	}
+}
```
It is sufficient to count only for passed and skipped. Filenames are needed only for failed.

kyu08 replied:
Fixed in 8d95a95e1a3b64b662f2a119d3e9e0aa703313f8.

kyu08 commented on reporter/reporter.go:
```diff
@@ -47,12 +47,17 @@ type Reporter interface {
 func Run(f func(r Reporter), opts ...Option) bool {
 	r := run(f, opts...)
+	r.printTestSummary()
+
```
🙏 
Fixed in 151da5ad3fdf94747b602d83616cbc09d5961fe6.

zoncoen commented on reporter/context.go:
```diff
@@ -38,6 +38,14 @@ func WithNoColor() Option {
+		ctx.enabledTestSummary = true
+		ctx.testSummary = newTestSummary(enabledColor)
+	}
+}
```
> I understand you mean testSummary struct should not have enabledColor field and testContext.noColor's value should be passed to testSummary#String as needed.

That's right. In other words, `reporter. WithNoColor` option should control all output.

> So I feel there is no point to change about this.

The new implementation is also acceptable. 👍

zoncoen reviewed (approved):
Thank you for taking the time and patience to bring this into the final shape!
LGTM 💯

kyu08 commented:
Thank you for reviewing my clumsy code using your precious time too!🙏
And I'm sorry that lack of understanding of this project of me leads increasing time and labor to review this PR.
I probably can deal with it more smoothly next time, so I am willing to contribute if there is an issue I can solve.

kyu08 commented:
Thank you for reviewing my pull request the other day. I'm looking forward to using the test summary feature, and I’d really appreciate it if you could roll out the next version when it’s feasible. 
Thank you.
